File: BraintreeModels/src/test/java/com/braintreepayments/api/CardConfigurationUnitTest.java
package com.braintreepayments.api;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static junit.framework.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
public class CardConfigurationUnitTest {

    @Test
    public void fromJson_parsesFullInput() throws JSONException {
        JSONObject input = new JSONObject()
                .put("collectDeviceData", true)
                .put("supportedCardTypes", new JSONArray()
                        .put("American Express")
                        .put("Discover")
                        .put("JCB")
                        .put("MasterCard")
                        .put("Visa"));
        CardConfiguration sut = CardConfiguration.fromJson(input);

        assertTrue(sut.isFraudDataCollectionEnabled());
        assertEquals(5, sut.getSupportedCardTypes().size());
        assertEquals("American Express", sut.getSupportedCardTypes().get(0));
        assertEquals("Discover", sut.getSupportedCardTypes().get(1));
        assertEquals("JCB", sut.getSupportedCardTypes().get(2));
        assertEquals("MasterCard", sut.getSupportedCardTypes().get(3));
        assertEquals("Visa", sut.getSupportedCardTypes().get(4));
    }

    @Test
    public void fromJson_whenInputNull_returnsConfigWithDefaultValues() {
        CardConfiguration sut = CardConfiguration.fromJson(null);
        assertFalse(sut.isFraudDataCollectionEnabled());
        assertEquals(0, sut.getSupportedCardTypes().size());
    }

    @Test
    public void fromJson_whenInputEmpty_returnsConfigWithDefaultValues() {
        CardConfiguration sut = CardConfiguration.fromJson(new JSONObject());
        assertFalse(sut.isFraudDataCollectionEnabled());
        assertEquals(0, sut.getSupportedCardTypes().size());
    }
}

File: BraintreeCore/src/test/java/com/braintreepayments/api/CardConfigurationUnitTest.kt
package com.braintreepayments.api

import org.robolectric.RobolectricTestRunner
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(RobolectricTestRunner::class)
class CardConfigurationUnitTest {
    @Test
    fun fromJson_parsesFullInput() {
        val input = JSONObject()
            .put("collectDeviceData", true)
            .put(
                "supportedCardTypes", JSONArray()
                    .put("American Express")
                    .put("Discover")
                    .put("JCB")
                    .put("MasterCard")
                    .put("Visa")
            )
        val (supportedCardTypes, isFraudDataCollectionEnabled) = CardConfiguration(input)
        assertTrue(isFraudDataCollectionEnabled)
        assertEquals(5, supportedCardTypes.size)
        assertEquals("American Express", supportedCardTypes[0])
        assertEquals("Discover", supportedCardTypes[1])
        assertEquals("JCB", supportedCardTypes[2])
        assertEquals("MasterCard", supportedCardTypes[3])
        assertEquals("Visa", supportedCardTypes[4])
    }

    @Test
    fun fromJson_whenInputNull_returnsConfigWithDefaultValues() {
        val (supportedCardTypes, isFraudDataCollectionEnabled) = CardConfiguration(null)
        assertFalse(isFraudDataCollectionEnabled)
        assertEquals(0, supportedCardTypes.size)
    }

    @Test
    fun fromJson_whenInputEmpty_returnsConfigWithDefaultValues() {
        val (supportedCardTypes, isFraudDataCollectionEnabled) = CardConfiguration(JSONObject())
        assertFalse(isFraudDataCollectionEnabled)
        assertEquals(0, supportedCardTypes.size)
    }
}
